src/Conversion/KrnlToAffine/KrnlToAffine.cpp

```diff
@@ -25,6 +25,215 @@
 using namespace mlir;
 
 namespace {
+// Since Krnl Dialect allows optimizations to be specified in the form of
```

Comment:
The explanation below is great, and I am trying to clarify it a bit more. If I look at the example given in permute:
```mlir
    %ii, %jj, %kk = krnl.define_loops 3
    // Blocking each loop by a factor of 4.
    %ib, %il = krnl.block %ii 4 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
    %jb, %jl = krnl.block %jj 4 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
    %kb, %kl = krnl.block %kk 4 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
    // Move iteration over tile coordinates to be the outer loops and iterateion over
    // the inter-tile elements to be the inner loops.
    krnl.permute(%ib, %il, %jb, %jl, %kb, %kl) [0, 3, 1, 4, 2, 5] : !krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop
    krnl.iterate(%ib, %il, %jb, %jl, %kb, %kl) with (%ii -> %i = 0 to 1024, %jj -> %j = 0 to 2048, %kk -> %k = 0 to 4096)  {
    }
```

there is a single iterate that list all of the blocked and intra-tile iterations `%ib, %il, %jb, %jl, %kb, %kl`. As an aside, the order in the iterate and permute input parameter has not been defined: what is it? Because ii came before jj in the define output, does it mean that the ib and il must be before the jb and jl in the permute input? Clarifying this would be helpful. I understand that the actual order is fixed by the permute literal `[0, 3, 1, 4, 2, 5]`.

Now if I compare this generic tiling to the example you give below:
```mlir
  %ii, %ij = krnl.define_loops 2
  %ib, %il = krnl.block %ii 5 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
  %jb, %jl = krnl.block %ij 4 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
  krnl.permute(%ib, %il, %jb, %jl) [0, 2, 1, 3] : !krnl.loop, !krnl.loop,
      !krnl.loop, !krnl.loop
  krnl.iterate(%ib, %jb) with (%ii -> %i = 0 to 10, %ij -> %j = 0 to 20) {
     %alloc = alloc() : memref<10 x f32>
     krnl.iterate(%il, %jl) with () {
        %foo = addi %i, %j : index
     }
     dealloc %alloc : memref<10 x f32>
  }
```

The first iterate contains only the two blocked iterations. This thus indicate that the alloc operation will be located directly inside of the two outer blocked iterations.  Then the two inner tile loop iterators are given with the add in the innermost. Thus the locations of the allocate/add are fully defined by the iterates and which loop iterators they entailed.

If my understanding is correct, that is good. I still wonder what the permute was about. To me, it is still redundant before, because we could have used the "order" in which the loop iterations are given as parameter to the `iterate` operation. Now it looks that since the original `%ib, %il, %jb, %jl` can be further separated into two distinct iterate operation, there is a second order of redundancy.

Ideally there would be no need for the permute in these cases; best would be that they become implicit (that is, if you need them,  they be automatically generated, as the order in the iterate would directly reflect the order of the loops. If that is too much work, we minimally need to be explicit that the order in the iterate is lexicographical (how they are listed in the define, permute outputs), and that the permute operation that reorder them must match the use across the iterate operation that may materialize them.

Reply:
To clarify when/how the order of appearance of loop handlers are meaningful:

- The order in which loop handlers are defined by `define_loops` operation is irrelevant and arbitrary. This ordering information is not used anywhere in any ways.

- The order in which loop handlers appear **after** `with` keyword stipulates the ordering of loops in the unoptimized case, and if an optimization recipe is present, optimizations will be applied successively to this unoptimized piece of code.

- The order in which loop handlers appear **before** `with` keyword **corresponds to, but does not determine** the ordering of loops after the original program has been optimized. So you will see the appearance of tiled loop handlers here. However, you can't put in arbitrary ordering of loop handlers here because there is only 1 single correct loop ordering, which corresponds to how loops are ordered in the optimized code after applying recipes to the unoptimized code. 

- The ordering of loop handlers within the input of krnl.permute operation is irrelevant and arbitrary. Since we only care about which new positions to send these loops to.

This is just the way it is right now, I don't think the third point makes total sense.


> The first iterate contains only the two blocked iterations. This thus indicate that the alloc operation will be located directly inside of the two outer blocked iterations. Then the two inner tile loop iterators are given with the add in the innermost. Thus the locations of the allocate/add are fully defined by the iterates and which loop iterators they entailed.

Your understanding is right.

> If my understanding is correct, that is good. I still wonder what the permute was about. To me, it is still redundant before, because we could have used the "order" in which the loop iterations are given as parameter to the iterate operation. 

I agree. Permute makes implementation a bit easier, so it's probably not their for the right reason. I agree that ultimately this redundancy should be eliminated by removing permute operation.

> Now it looks that since the original %ib, %il, %jb, %jl can be further separated into two distinct iterate operation, there is a second order of redundancy.

Why is there a second order of redundancy?

> Ideally there would be no need for the permute in these cases; best would be that they become implicit (that is, if you need them, they be automatically generated, as the order in the iterate would directly reflect the order of the loops. 

This is a worthwhile goal to work towards.

> If that is too much work, we minimally need to be explicit that the order in the iterate is lexicographical (how they are listed in the define, permute outputs), and that the permute operation that reorder them must match the use across the iterate operation that may materialize them.

Yes, they are and should be lexicographical in iterate op. However, they don't need to correspond to how they're listed in define/permute simply because of the way define and permuate are interpreted (define create loops, and says nothing about the ordering; permute argument specifies the permuted loop ordering, so don't care about how inputs are ordered). Please let me know if this makes sense or not.